src/java.base/share/classes/jdk/internal/vm/VThreadSummary.java

```diff
@@ -0,0 +1,174 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.vm;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.misc.Unsafe;
+import sun.nio.ch.Poller;
+
+/**
+ * The implementation for the jcmd Thread.vthread_summary diagnostic command.
+ */
+public class VThreadSummary {
+    private static final Unsafe U = Unsafe.getUnsafe();
+    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
+
+    // maximum number of thread containers to print
+    private static final int MAX_THREAD_CONTAINERS = 256;
+
+    private VThreadSummary() { }
+
+    /**
+     * Invoked by the VM to print virtual thread summary information.
+     * @return the UTF-8 encoded information to print
+     */
+    private static byte[] print() {
+        StringBuilder sb = new StringBuilder();
+
+        // print virtual thread scheduler
+        printSchedulers(sb);
+        sb.append(System.lineSeparator());
+
+        // print I/O pollers if initialized
+        if (!U.shouldBeInitialized(Poller.class)) {
+            printPollers(sb);
+            sb.append(System.lineSeparator());
+        }
+
+        return sb.toString().getBytes(StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Print information on the virtual thread schedulers to given string builder.
+     */
+    private static void printSchedulers(StringBuilder sb) {
+        sb.append("Virtual thread scheduler:")
+                .append(System.lineSeparator());
+        sb.append(JLA.virtualThreadDefaultScheduler())
+                .append(System.lineSeparator());
+
+        sb.append(System.lineSeparator());
+
+        sb.append("Timeout schedulers:")
+                .append(System.lineSeparator());
+        var schedulers = JLA.virtualThreadDelayedTaskSchedulers().toList();
+        for (int i = 0; i < schedulers.size(); i++) {
+            sb.append('[')
+                    .append(i)
+                    .append("] ")
+                    .append(schedulers.get(i))
+                    .append(System.lineSeparator());
+        }
+    }
+
+    /**
+     * Print information on threads registered for I/O to the given string builder.
+     */
+    private static void printPollers(StringBuilder sb) {
+        Poller masterPoller = Poller.masterPoller();
+        List<Poller> readPollers = Poller.readPollers();
+        List<Poller> writePollers = Poller.writePollers();
+
+        if (masterPoller != null) {
+            sb.append("Master I/O poller:")
+                    .append(System.lineSeparator())
+                    .append(masterPoller)
+                    .append(System.lineSeparator());
+            sb.append(System.lineSeparator());
```

Comment:
Is this style trying to draw attention to the blank lines between sections? It looks odd to me to not chain the final append as well.

Reply:
It is a blank line, shouldn't be chained to the previous statement.